paperparcel-kotlin/src/main/java/paperparcel/PaperParcelable.kt

```diff
@@ -11,38 +11,38 @@ import java.lang.reflect.Method
  * @since 2.0
  */
 interface PaperParcelable : Parcelable {
-  private object Cache {
-    val writeMethods = mutableMapOf<Class<*>, Method>()
-  }
-
   override fun describeContents() = 0
 
   override fun writeToParcel(dest: Parcel, flags: Int) {
-    if (!Cache.writeMethods.containsKey(javaClass)) {
-      val annotatedClass = findAnnotatedClass(javaClass)
-      val generatedClass = Class.forName(implementationName(annotatedClass))
-      val args = arrayOf(annotatedClass, Parcel::class.java, Int::class.javaPrimitiveType)
-      val writeMethod = generatedClass.getDeclaredMethod("writeToParcel", *args)
-      writeMethod.isAccessible = true
-      Cache.writeMethods.put(javaClass, writeMethod)
+    val writeMethod: Method = synchronized(writeMethods) {
+      writeMethods.getOrPut(javaClass) {
+        val annotatedClass = javaClass.annotatedClass
+        val generatedClass = Class.forName(annotatedClass.implementationName)
+        generatedClass
+                .getDeclaredMethod("writeToParcel", annotatedClass, Parcel::class.java, Int::class.javaPrimitiveType)
+                .apply {
+                  isAccessible = true
+                }
+      }
     }
-    Cache.writeMethods[javaClass]!!.invoke(null, this, dest, flags)
+    writeMethod.invoke(null, this, dest, flags)
   }
 
-  private fun findAnnotatedClass(type: Class<*>): Class<*> {
-    if (type.isAnnotationPresent(PaperParcel::class.java)) {
-      return type
-    }
-    if (type == Any::class.java) {
-      throw IllegalArgumentException(
-          "Cannot find @${PaperParcel::class.java.simpleName} on ${javaClass.name}.")
+  private val Class<*>.annotatedClass: Class<*>
+    get() {
+      if (isAnnotationPresent(PaperParcel::class.java)) {
+        return this
+      }
+      require(this != Any::class.java) {
+        "Cannot find @${PaperParcel::class.java.simpleName} on ${this@PaperParcelable.javaClass.name}."
+      }
+      return superclass.annotatedClass
     }
-    return findAnnotatedClass(type.superclass)
-  }
 
-  private fun implementationName(type: Class<*>): String {
-    val packageName = type.`package`.name
-    val className = type.name.substring(packageName.length + 1).replace('$', '_')
-    return packageName + ".PaperParcel" + className
+  private val Class<*>.implementationName: String
+    get() = "${`package`.name}.PaperParcel${name.substring(`package`.name.length + 1).replace('$', '_')}"
+
+  private companion object {
+    val writeMethods = HashMap<Class<*>, Method>()
```

Comment:
I changed this to `HashMap` rather than `mutableMapOf` which uses a `LinkedHashMap` because `LinkedHashMap` uses more memory and unless you iterate over the map's contents there is no different in behaviour.

Reply:
👌